src/hugr/views.rs

```diff
@@ -303,8 +300,59 @@ pub trait HugrView: sealed::HugrInternals {
     }
 }
 
+/// Trait for views that provides a guaranteed bound on the type of the root node.
+pub trait RootTagged: HugrView {
+    /// The kind of handle that can be used to refer to the root node.
+    ///
+    /// The handle is guaranteed to be able to contain the operation returned by
+    /// [`HugrView::root_type`].
+    type RootHandle: NodeHandle;
+}
+
+/// A view of the whole Hugr.
+/// (Just provides static checking of the type of the root node)
+pub struct RootChecked<H, Root = Node>(H, PhantomData<Root>);
+
+impl<H: HugrView, Root: NodeHandle> RootChecked<H, Root> {
+    /// Create a hierarchical view of a whole HUGR
+    ///
+    /// # Errors
+    /// Returns [`HugrError::InvalidTag`] if the root isn't a node of the required [`OpTag`]
+    pub fn try_new(hugr: H) -> Result<Self, HugrError> {
+        check_tag::<Root>(&hugr, hugr.root())?;
+        Ok(Self(hugr, PhantomData))
+    }
+}
+
+impl<Root> RootChecked<Hugr, Root> {
+    /// Extracts the underlying (owned) Hugr
+    pub fn into_hugr(self) -> Hugr {
+        self.0
+    }
+}
+
+impl<H: AsRef<Hugr>, Root: NodeHandle> RootTagged for RootChecked<H, Root> {
+    type RootHandle = Root;
+}
+
+// Note do not implement AsMut<Hugr> - that would get us the `impl HugrMutInternals`
```

Comment:
Note this contrary to your initial idea @aborgna-q. If you try to impl AsMut<Hugr>, you'll have to remove the impl of HugrMutInternals (and HugrMut) here, and then the test will fail. The test may be enough to justify removing this comment, possibly.

Reply:
Uhm, this is an easy pit to fall in.

Since the auto-impl for `HugrMutInternals` requires `RootTagged`, why not check the root type before dereferencing with `as_mut` there?
The generic constraints in `impl<T: RootTagged + AsMut<Hugr>> HugrMutInts for T` kind of tell you that you're adding the extra check on top of the reference.

Reply:
Hmmm, the answer might be - premature optimization, i.e. that the `impl HugrMutInternals` for `T: AsMut<Hugr>` then gets to skip the check....so ok, let's not prematurely optimize, SGTM ;)

Note there was an unexpected subtle consequence that I had to rule out in 07e4f1097cf936d8327881bdc0e2b118ba627613

Reply:
(and this does let anyone call `.as_mut().replace_op(...)` and bypass the check, which I'm not very keen on....)

Reply:
So overall, I have done as you say, but I'm not sure why this is better, TBH

Reply:
See 2be330dc7a16dccc7dacbf04bbe84c63fa5e283c, which HEAD reverts - I actually think I prefer the former; it needs an extra `impl HugrMut` (but, empty!) so that seems minimally harmful to me